parlai/utils/torch.py

```diff
@@ -235,3 +238,229 @@ def forward(self, xs):
+        elif isinstance(item, dict):
+            if item == {}:
+                # terrible edge case: the empty dict. return an infinite list
+                # of empty dicts and we'll figure out its correct size later
+                return itertools.repeat({})
+            # we start with Dict[key,tensor]
+            # we map it to d: Dict[key, List[Tensor]], where we have split each mapping
+            d = {k: PipelineHelper.split(v, split_size, dim) for k, v in item.items()}
+            # now we transpose it and return List[Dict[key, Tensor]]
+            return [
+                dict(zip(d.keys(), values))  # type: ignore
+                for values in zip(*(d[k] for k in d.keys()))
+            ]
+        else:
+            raise TypeError(f"Cannot split type {type(item)}")
+
+    @staticmethod
+    def join(items: List[Chunk], dim=0) -> Chunk:
+        """
+        Join chunks back together, the inverse of split.
+
+        :param items:
+            All the output chunks. Each chunk may be a tensor or a group of
+            tensors.
+        :param dim:
+            The dimension to join along.
+        """
+        if len(items) == 0:
+            raise IndexError("Cannot rejoin an empty list of chunks.")
+        item0 = items[0]
+        if isinstance(item0, torch.Tensor):
+            # base case
+            return torch.cat(items, dim=dim)  # type: ignore
+        elif isinstance(item0, tuple):
+            return tuple(PipelineHelper.join(x, dim=dim) for x in zip(*items))  # type: ignore
+        elif isinstance(item0, dict):
+            keys = item0.keys()
+            return {  # type: ignore
+                k: PipelineHelper.join([c[k] for c in items], dim=dim)  # type: ignore
+                for k in keys
+            }
+        else:
+            raise TypeError(f'Cannot join list of type {type(item0)}')
+
+    @staticmethod
+    def layer_assignment(
+        layer_no: int, num_layers: int, num_gpus: Optional[int] = None
+    ) -> str:
+        """
+        Determine which device a layer should be on.
+
+        :param layer_no:
+            0-indexed layer number
+        :param num_layers:
+            Total number of layers to parallelize
+        :param num_gpus:
+            Number of gpus to distribute over. If None, use all available devices.
+        :returns:
+            A specific device, e.g. "cuda:0" or "cuda:3".
+        """
+        if num_gpus is None:
+            num_gpus = torch.cuda.device_count()  # type: ignore
+        assert isinstance(num_gpus, int)
+        # device_no = layer_no // int(math.ceil(num_layers / num_gpus))
+        # return f'cuda:{device_no}'
+        if layer_no == 0:
+            return 'cuda:0'
+        elif num_layers < num_gpus:
+            return f'cuda:{layer_no}'
+        else:
+            assert isinstance(num_gpus, int)
+            device_no = 1 + (layer_no - 1) // int(
+                math.ceil((num_layers - 1) / (num_gpus - 1))
```

Comment:
this is a bit confusing to me
say you have 12 layers and 2 gpus.
then layer 0 --> gpu 0 
layer n > 0 --> gpu 1

is this how it is supposed to work?

(also you don't need the cast to int if you are doing a math.ceil)